### PR TITLE
Added image model validation

### LINKED ISSUE #4: Image URL Validation
#### As a user I want to be prevented from submitting invalid image URLs.

Acceptance criteria:
- [x] I cannot successfully save an image with an invalid URL.
- [x] An error message is associated with the appropriate input field on
  failure.

Discussion Topic:
- [x] What is a valid image URL?

Dependencies:
- Save Image Link

### PR DESCRIPTION
**Pull request by wpliao1989**
Closes #4 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

**Comment by wpliao1989** [comment]
I added url validation, could you take a look at this when you have a minute? @XanderStrike

File: app/models/image.rb [hunk 1]
```diff
@@ -1,2 +1,16 @@
 class Image < ApplicationRecord
+  validates :url, presence: true
+  validate :url_is_valid
+
+  private
+
+  def url_is_valid
+    URI.parse(url)
+    response = HTTParty.head(url)
```
Comment (XanderStrike) [inline app/models/image.rb]: This is the most interesting way of checking the validity of a URL I've seen 😉 

That said, it's not an efficient way to do it. You're making the user wait while a request goes through, and it could falsely fail for invalid URLs because of temporary network issues. Plus every time the app goes through CI it'll need to hit some test URL to make sure this works (or you'll have to stub it out, which defeats the purpose of the test).

ActiveRecord provides a neat `format` helper that can assert the format of a field against a regular expression. Check out the [docs](http://guides.rubyonrails.org/active_record_validations.html#format) and try that route.

**Comment by XanderStrike** [comment]
I noticed there were no tests in this PR whatsoever. I'm gonna need to see full test coverage before I can approve this pull.

**Comment by wpliao1989** [comment]
Thanks for the input! Yes I was looking for an alternative for validating the url, so a unit test can be written, otherwise I have to use mock or something.

**Comment by wpliao1989** [comment]
Updated

File: app/models/image.rb [hunk 2]
```diff
@@ -1,2 +1,4 @@
 class Image < ApplicationRecord
+  validates :url, presence: { message: "Url can't be blank" }
+  validates :url, format: { with: URI.regexp(%w(http https)), message: 'Url is invalid' }
```
Comment (XanderStrike) [inline app/models/image.rb]: `URI.regexp` is designed for pulling URLs out of text blocks (like in this comment, note how http://asdfadsfasdf.fasdfsdf is automatically linked, that's what it does).

What you've got here is actually more permissive even than what Github is doing. For example, `http://` will pass. Try googling around for a good URL validation regular expression. 

You might also want to ensure they're image files by requiring that they end with `[.png|.jpg|.gif]` or something.

File: test/models/image_test.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def test_url_scheme_not_http_https
+    image = Image.new(url: 'ftp://www.google.com')
+    refute image.valid?
+    assert_equal 'Url is invalid', image.errors.messages[:url].first
+  end
+
+  def test_url_with_weird_chars
+    image = Image.new(url: 'http://उदाहरण.परीक्ष')
```
Comment (XanderStrike) [inline test/models/image_test.rb]: These characters aren't actually valid in the current URI spec ([here](http://www.ietf.org/rfc/rfc3986.txt) if you want to read but it's dense), so you should actually be refuting that it's valid (and then changing your validation so the test passes again)

**Comment by XanderStrike** [comment]
Please slap on some tests for the controller as well, since you've added some logic as part of this PR.

**Comment by wpliao1989** [comment]
Got it, thanks!

File: app/controllers/images_controller.rb [hunk 4]
```diff
@@ -4,12 +4,18 @@ def new
   end
 
   def create
-    @image = Image.create!(image_params)
-    redirect_to @image, notice: 'New image created!'
+    @image = Image.new(image_params)
+    respond_to do |format|
```
Comment (bboe) [inline app/controllers/images_controller.rb]: Is `respond_to` required for what we're doing?

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -4,12 +4,18 @@ def new
...
+      if @image.save
+        format.html { redirect_to @image, notice: 'New image created!' }
+      else
+        format.html { render action: :new }
```
Comment (bboe) [inline app/controllers/images_controller.rb]: What status code does this result in? Is it the correct one?

File: app/models/image.rb (hunk 2 above)
Comment (bboe) [inline app/models/image.rb]: Did you have any questions about Alex's suggestion?

File: test/models/image_test.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  def test_url_http_valid
+    image = Image.new(url: 'http://www.google.com')
+    assert image.valid?
+  end
+
+  def test_url_https_valid
```
Comment (bboe) [inline test/models/image_test.rb]: Can we combine these assertions into a single `url_valid` test? Are there more valid formats that we should ensure remain valid?

File: test/models/image_test.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    image = Image.new(url: 'https://www.google.com')
+    assert image.valid?
+  end
+
+  def test_url_must_be_present
```
Comment (bboe) [inline test/models/image_test.rb]: Is there another value, other than `''`, that when provided will result in the same message?

File: test/models/image_test.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    image = Image.new(url: '')
+    refute image.valid?
```
Comment (bboe) [inline test/models/image_test.rb]: Consider using `refute_predicate`. `refute` and `assert` do not provide useful context when they fail. `refute_predicate` and `assert_predicate` do.

File: test/models/image_test.rb [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    assert_equal "Url can't be blank", image.errors.messages[:url].first
+  end
+
+  def test_url_must_have_scheme
+    image = Image.new(url: 'www.google.com')
+    refute image.valid?
+    assert_equal 'Url is invalid', image.errors.messages[:url].first
+  end
+
+  def test_url_scheme_not_http_https
```
Comment (bboe) [inline test/models/image_test.rb]: Can we combine all the `url is invalid` assertions into a single test? Doing so will keep you from repeating a bunch of the same code, as when you update the REGEX we will end up with a significant number of assertions.

File: test/models/image_test.rb [hunk 10]
```diff
@@ -0,0 +1,36 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  def test_url_http_valid
+    image = Image.new(url: 'http://www.google.com')
+    assert image.valid?
```
Comment (bboe) [inline test/models/image_test.rb]: Consider `assert_predicate`.

File: test/models/image_test.rb (hunk 3 above)
Comment (bboe) [inline test/models/image_test.rb]: Regardless of if this is valid or not (sounds like it's not) you could add this as a single case in either your valid test, or your invalid test.

**Comment by bboe** [comment]
@wpliao1989 Could you please address some of the comments. Overall the change is a great first step.

One other thing to consider. Now that URL is a required field, is there anything we can change in the database to help enforce that constraint?

File: app/models/image.rb (hunk 2 above)
Comment (wpliao1989) [inline app/models/image.rb]: I agree that a different regex should be used but I think I'll skip checking file extension because    it doesn't tell any fact about the file format.

**Comment by wpliao1989** [comment]
Updated

File: app/controllers/images_controller.rb [hunk 11]
```diff
@@ -4,12 +4,16 @@ def new
   end
 
   def create
-    @image = Image.create!(image_params)
-    redirect_to @image, notice: 'New image created!'
+    @image = Image.new(image_params)
+    if @image.save
+      redirect_to @image, notice: 'New image created!'
+    else
+      render action: :new, status: :bad_request
```
Comment (bboe) [inline app/controllers/images_controller.rb]: Great. `bad_request` is the one I originally chose as well. However, we generally use `unprocessible_entity` for failed forms.

File: app/models/image.rb [hunk 12]
```diff
@@ -1,2 +1,9 @@
 class Image < ApplicationRecord
+  validates :url, presence: { message: "Url can't be blank" }
```
Comment (bboe) [inline app/models/image.rb]: Do we need to manually add the message?

File: app/models/image.rb [hunk 13]
```diff
@@ -1,2 +1,9 @@
...
+  validates :url, format: { with: %r{
+  ^(http|https):
+\/\/[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.
+[a-z]{2,5}(([0-9]{1,5})?\/.
+*)?$
+}x, message: 'Url is invalid' }
```
Comment (bboe) [inline app/models/image.rb]: Is this message consistent with the rails default error messages?

File: test/controllers/images_controller_test.rb [hunk 14]
```diff
@@ -7,13 +7,19 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
   end
 
-  test 'should create' do
-    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
-    assert_difference('Image.count') do
-      post images_url image: { url: url }
-    end
-    assert_redirected_to image_url(Image.last)
-    assert_equal 'New image created!', flash[:notice]
+  test 'should create with valid url' do
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: I hope I didn't mislead you before. As far as the controller tests go we only need to test one success case, and one case for each failure that results from code in the controller. All other cases should be handled with the model tests.

Also for controller tests, we can still have multiple assertions, however, we should avoid having more than one GET/POST/DELETE (http request) per test. Note that in Rails5 this isn't really a hard requirement anymore, but it's still a good practice to separate different requests into their own methods.

That being said, I do like the way you split out a `test_valid_url` function. If this were the way to do things it is very clean.

File: test/controllers/images_controller_test.rb [hunk 15]
```diff
@@ -27,4 +33,21 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    end
+    assert_redirected_to image_url(Image.last)
+    assert_equal 'New image created!', flash[:notice]
+  end
+
+  def test_invalid_url(url, message = 'Url is invalid')
+    assert_no_difference('Image.count') do
+      post images_url image: { url: url }
+    end
+    assert_select '.error', message
```
Comment (bboe) [inline test/controllers/images_controller_test.rb]: What is the http response status?

File: test/models/image_test.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+class ImageTest < ActiveSupport::TestCase
+  def test_url_valid
+    assert_predicate Image.new(url: 'http://www.google.com'), :valid?
+    assert_predicate Image.new(url: 'https://www.google.com'), :valid?
+  end
+
+  def test_url_empty
+    message = "Url can't be blank"
+    test_url_invalid_helper Image.new(url: ''), message
+    test_url_invalid_helper Image.new(url: nil), message
```
Comment (bboe) [inline test/models/image_test.rb]: ⭐  good.

File: app/models/image.rb [hunk 17]
```diff
@@ -1,2 +1,9 @@
 class Image < ApplicationRecord
+  validates :url, presence: { message: "Url can't be blank" }
+  validates :url, format: { with: %r{
+  ^(http|https):
```
Comment (bboe) [inline app/models/image.rb]: Now that we have a regex, we need to test every class of success or failure including those success/fail edges where they exist. This is to ensure that someone who wants to update this regex in the future doesn't accidentally break something.

**Comment by bboe** [comment]
Great progress. There's a bit more work yet.

File: app/models/image.rb (hunk 12 above)
Comment (wpliao1989) [inline app/models/image.rb]: If not, the default message is "can't be blank". That would work too but I wanted to be more precise.

File: app/controllers/images_controller.rb (hunk 11 above)
Comment (wpliao1989) [inline app/controllers/images_controller.rb]: Got it.

File: test/controllers/images_controller_test.rb (hunk 14 above)
Comment (wpliao1989) [inline test/controllers/images_controller_test.rb]: Makes sense, working on it.

File: app/models/image.rb (hunk 12 above)
Comment (bboe) [inline app/models/image.rb]: When using the form helpers the error message should appear adjacent to the field itself so it should be pretty clear to the user.

Comment (wpliao1989) [inline app/models/image.rb]: Cool, yeah that would work.

**Comment by wpliao1989** [comment]
Updated

**Comment by wpliao1989** [comment]
Wait let me squash commits

**Comment by wpliao1989** [comment]
Updated

File: app/models/image.rb [hunk 18]
```diff
@@ -1,2 +1,48 @@
 class Image < ApplicationRecord
+  validates :url, presence: true
+  validates :url, format: { with: %r{
+    \A
+
+    # protocol identifier
+    (?:(?:https?|ftp)://)
```
Comment (bboe) [inline app/models/image.rb]: Do FTP urls work for loading images in most browsers?

File: app/models/image.rb [hunk 19]
```diff
@@ -1,2 +1,48 @@
...
+
+    # user:pass authentication
+    (?:\S+(?::\S*)?@)?
+
+    (?:
+      # IP address exclusion
+      # private & local networks
```
Comment (bboe) [inline app/models/image.rb]: Do you require this level of checking? We don't necessarily have to accept all possible image urls, so I think it's safe to say assume there is a domain name.

File: test/models/image_test.rb [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+      'http://foo.com/blah_(wikipedia)#cite-1',
+      'http://foo.com/blah_(wikipedia)_blah#cite-1',
+      'http://foo.com/unicode_(✪)_in_parens',
+      'http://foo.com/(something)?after=parens',
+      'http://☺.damowmow.com/',
+      'http://code.google.com/events/#&product=browser',
+      'http://j.mp',
+      'ftp://foo.bar/baz',
+      'http://foo.bar/?q=Test%20URL-encoded%20stuff',
+      'http://مثال.إختبار',
```
Comment (bboe) [inline test/models/image_test.rb]: Are these actually valid domain names?

File: test/models/image_test.rb [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+  def test_url_invalid_helper(image, message = 'is invalid')
+    refute_predicate image, :valid?
+    assert_equal message, image.errors.messages[:url].first
+  end
+
+  # rubocop:disable MethodLength
+  def valid_urls
+    [
+      'http://✪df.ws/123',
+      'http://userid:password@example.com:8080',
```
Comment (bboe) [inline test/models/image_test.rb]: Your port number regex component is: `(?::\d{2,5})?`. That means there should be a failing test for a 1-digit port and 6-digit port (failing edge cases) as well as the success edge cases 2-digit port, and 5-digit port. Additionally, since you allow only numbers as the port, you should also have a failing test if you try to use a letter in a place that would otherwise pass if it were a number.

These same edge cases should be everywhere there is a length limit (tld appears to have a minimum). Hence why I think your work will be much simpler if you don't include IP addresses (lots of edge cases).

**Comment by bboe** [comment]
These are a great start to enumerating the test cases. I think there can be some simplification and then a few more test cases to ensure the REGEX is defensively tested, that is, tests will fail if the regex is modified at all.

**Comment by bboe** [comment]
One more thing- a failing submission should be added to your flow test. That is the user first inputs an invalid URL, then you can verify the error message, and then they fix the error and submit.

**Comment by wpliao1989** [comment]
What's a flow test? Is it a controller test or a full integration test like Selenium tests?

**Comment by bboe** [comment]
> What's a flow test?

A flow test should represent the flow a user would take when using your website. For instance a new user may visit the home page, click on the new image button, input the form incorrectly, fix their input, and then observe the image they uploaded. At various steps in this flow you'll want to make assertions about what is on the page (e.g., the error message).

Your flow tests should be added to test/flows/.

**Comment by wpliao1989** [comment]
Updated

File: app/models/image.rb [hunk 22]
```diff
@@ -1,2 +1,9 @@
 class Image < ApplicationRecord
+  validates :url, presence: true
+  validates :url, format: { with: %r{
+  \A(http|https):\/\/
+[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\.
```
Comment (bboe) [inline app/models/image.rb]: `[\-\.]{1}` can be simplified with `(-|\.)`.

Comment (wpliao1989) [inline app/models/image.rb]: You're right! Good catch.

**Comment by wpliao1989** [comment]
Updated

File: db/migrate/20160726015238_add_url_constraint.rb [hunk 23]
```diff
@@ -0,0 +1,5 @@
+class AddUrlConstraint < ActiveRecord::Migration[5.0]
+  def change
+    change_column_null(:images, :url, false)
```
Comment (bboe) [inline db/migrate/20160726015238_add_url_constraint.rb]: ⭐

File: test/flows/image_flows_test.rb [hunk 24]
```diff
@@ -0,0 +1,17 @@
+require 'flow_test_helper'
+
+class ImageFlowsTest < FlowTestCase
+  test 'create a new image flow' do
+    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+
+    visit new_image_path
```
Comment (bboe) [inline test/flows/image_flows_test.rb]: Would your users go directly to this path? Or might they go to the root path and then click on a button to get them to the new page (I don't know if you have such a link).

File: test/flows/image_flows_test.rb [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+
+    fill_in('Url', with: 'test')
+    click_button('Create Image')
+    assert_selector('.error', text: 'is invalid')
```
Comment (bboe) [inline test/flows/image_flows_test.rb]: Can we make an assertion about the path at this point to ensure we're on the right page?

File: test/flows/image_flows_test.rb [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+
+    fill_in('Url', with: url)
+    click_button('Create Image')
+    assert page.has_content?('New image created!')
```
Comment (bboe) [inline test/flows/image_flows_test.rb]: Can we make an assertion about the path here? The nice thing about path assertions (if you're using the right one) is it ensures that the page is loaded prior to making other assertions on the page.

File: test/models/image_test.rb [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+      'http://foo.com/(something)?after=parens',
+      'http://code.google.com/events/#&product=browser',
+      'http://j.mp',
+      'http://foo.bar/?q=Test%20URL-encoded%20stuff'
+    ]
+  end
+
+  def invalid_urls
+    [
+      'http://',
```
Comment (bboe) [inline test/models/image_test.rb]: What about a URL that starts with a newline or other whitespace? Similarly one that ends with whitespace?

Comment (bboe) [inline test/models/image_test.rb]: What about `http//a.com`? E.g., ensuring the `:` exists.

Comment (bboe) [inline test/models/image_test.rb]: Adding a list of other things to check for:

* `http:/a.com`

File: test/models/image_test.rb [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+
+  def test_url_invalid_helper(image, message = 'is invalid')
+    refute_predicate image, :valid?
+    assert_equal message, image.errors.messages[:url].first
+  end
+
+  # rubocop:disable MethodLength
+  def valid_urls
+    [
+      'http://www.google.co',
```
Comment (bboe) [inline test/models/image_test.rb]: Your regex permits numbers in the domain and subdomain portions. I don't see any valid examples with that here.

Similarly, I don't see

* any CAPITAL letters in each text-permitting section
* more than one subdomain
* domain and subdomain with a hyphen
* Multiple consecutive slashes (`http://a.com///`)

Also, is port 0 allowed?

File: test/models/image_test.rb [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+      'http://?',
+      'http://?/',
+      'http://#',
+      'http://#/',
+      'http://foo.bar?q=Spaces should be encoded',
+      '//',
+      '//a',
+      '///a',
+      '///',
+      'http:///a',
```
Comment (bboe) [inline test/models/image_test.rb]: This is not valid for more than one reason. If's a good idea to ensure they fail due to exactly one reason (e.g., 3 slashes).

File: test/models/image_test.rb [hunk 30]
```diff
@@ -0,0 +1,101 @@
...
+      'http://⌘.ws/',
+      'http://مثال.إختبار',
+      'http://例子.测试',
+      'http://☺.damowmow.com/',
+      'http://www.google.c',
+      'http://www.google.comeon',
+      'http://www.google.com:',
+      'http://www.google.com:123456'
+    ]
+  end
```
Comment (bboe) [inline test/models/image_test.rb]: ⭐  This is a great list. Good work.

**Comment by bboe** [comment]
Looking a lot more complete.

File: test/flows/image_flows_test.rb (hunk 24 above)
Comment (wpliao1989) [inline test/flows/image_flows_test.rb]: I don't have one... yet.

File: test/models/image_test.rb (hunk 28 above)
Comment (wpliao1989) [inline test/models/image_test.rb]: Port zero is allowed, yes.

Comment (wpliao1989) [inline test/models/image_test.rb]: Great test cases

**Comment by wpliao1989** [comment]
Updated

**Comment by wpliao1989** [comment]
Forgot about flow test. Working on that.

**Comment by wpliao1989** [comment]
Updated

File: test/flows/image_flows_test.rb (hunk 24 above)
Comment (bboe) [inline test/flows/image_flows_test.rb]: That's fine. Just be sure to update it later when you do.

File: test/flows/image_flows_test.rb [hunk 31]
```diff
@@ -0,0 +1,19 @@
...
+class ImageFlowsTest < FlowTestCase
+  test 'create a new image flow' do
+    url = 'http://www.appfolio.com/images/html/apm-logo-v2.png'
+
+    visit new_image_path
+
+    fill_in('Url', with: 'test')
+    click_button('Create Image')
+    assert_selector('.error', text: 'is invalid')
+    assert_equal images_path, current_path
```
Comment (bboe) [inline test/flows/image_flows_test.rb]: In the order that you've done this I think this is okay assuming `assert_selector` is a retry until selector is available type assertion.

For completeness. `assert_equal` performs that check against the view immediately. When using capybara backed by selenium which actually uses a web browser, there is a race condition after the button click between how fast the test code says `assert_equal` and how fast the browser actually updates the page. If `assert_equal` runs first, the browser may still have the form's URL in it's path. As a result it's often a good idea to use `has_current_path?`.

File: test/flows/image_flows_test.rb [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+
+    fill_in('Url', with: url)
+    click_button('Create Image')
+    assert_equal image_path(Image.last), current_path
```
Comment (bboe) [inline test/flows/image_flows_test.rb]: This one is definitely a race condition.

File: test/models/image_test.rb [hunk 33]
```diff
@@ -0,0 +1,116 @@
...
+      'http://例子.测试',
+      'http://☺.damowmow.com/',
+      'http://www.google.c',
+      'http://www.google.comeon',
+      'http://www.google.com:',
+      'http://www.google.com:123456',
+      'http:/a.com',
+      ' http://www.google.com',
+      '\nhttp://www.google.com',
+      'http://www.google.com '
```
Comment (bboe) [inline test/models/image_test.rb]: Looks like `http//a.com` or similar is still missing.

**Comment by bboe** [comment]
Nearly perfect. These changes should be pretty straightforward, so when those are done go ahead and squash them to their atomic commits (just one I think) and then merge. 👍

**Comment by bboe** [comment]
Actually, it looks like there is an issue with the server: https://wpliao1989s-pr-20.herokuapp.com/

Any idea why that's failing?

**Comment by wpliao1989** [comment]
The review app broke because when it was deployed it didn't have db:migrate as post deploy script. It's fixed now.

**Comment by wpliao1989** [comment]
Updated

**Comment by bboe** [comment]
👍 good to merge
